samples/client/petstore/python-experimental/petstore_api/models/additional_properties_any_type.py

```diff
@@ -64,7 +64,7 @@ class AdditionalPropertiesAnyType(ModelNormal):
     validations = {
     }
 
-    additional_properties_type = (bool, date, datetime, dict, float, int, list, str,)  # noqa: E501
+    additional_properties_type = ({str: (bool, date, datetime, dict, float, int, list, str, none_type)},)  # noqa: E501
```

Comment:
So in python all properties in an object type model are stored in a dictionary named _data_store.
You can see our assignment to it [here](https://github.com/spacether/openapi-generator/blob/master/samples/client/petstore/python-experimental/petstore_api/model_utils.py#L98).

We require that key variable names are strings [here](https://github.com/spacether/openapi-generator/blob/master/samples/client/petstore/python-experimental/petstore_api/model_utils.py#L68).
The intention in defining these additional_properties_type is describing the value types that may be assigned.
So if the object spec does not include a property "cat",
we can assign the property "cat" which is the key. That assignment is handed in the [`__setattr__`](https://github.com/spacether/openapi-generator/blob/master/samples/client/petstore/python-experimental/petstore_api/model_utils.py#L174) and [`set_attribute`](https://github.com/spacether/openapi-generator/blob/master/samples/client/petstore/python-experimental/petstore_api/model_utils.py#L49) methods, which require that "cat" be a string.
We use additional_properties_type to describe the type values only.
Because we already have the key string portion taken care of [here](https://github.com/spacether/openapi-generator/blob/master/samples/client/petstore/python-experimental/petstore_api/model_utils.py#L68)
additional_properties_type should only have the value `(bool, date, datetime, dict, float, int, list, str,)`
Can you update your code to do that?
If you have a different interpretation of what should be happening here can you explain it?

Reply:
Should we handle this differently if {} anyType is defining additionalProperties on object type schema vs being in a object type schema's CodegenProperty?
1. in an object type schema, the place that we store variables already exists and is `_data_store`, we are only defining value types here, so I think that we should use:
`(bool, date, datetime, dict, float, int, list, str,)`
2. in a property, we are assigning a value. Should this data type be:
  a. `(bool, date, datetime, dict, float, int, list, str,)` or
  b `{str: (bool, date, datetime, dict, float, int, list, str,)}`
It sounds like for the CodegenProperty case we should be using definition b, is that correct?

Reply:
Another possible solution here is to keep:

`additional_properties_type = ({str: (bool, date, datetime, dict, float, int, list, str, none_type)},)`
as-is in both additional_properties_type and model properties.
If we did that we would need to:

- update model docstring [here](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_templates/docstring_openapi_validations.mustache#L6) to describe the updated additional_properties_type and how to use it
- update the set_attribute method [here](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_templates/method_set_attribute.mustache#L18) to ignore the `str` key in additional_properties_type and use the dict values only which are `(bool, date, datetime, dict, float, int, list, str, none_type)`

If we didd that, our documentation [here](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/python/python-experimental/model_doc.mustache#L22):
Would produce descriptions like:
```
# additional_properties_any_type.AdditionalPropertiesAnyType

## Properties
Name | Type | Description | Notes
------------ | ------------- | ------------- | -------------
**name** | **str** |  | [optional] 
**any string name** | **{str: (bool, date, datetime, dict, float, int, list, str, none_type)}** | any string name can be used but the value must be the correct type | [optional]
```
That is concerning to me because we are listing that an additionalProperty can have a name of any string, but it looks like the value should be a dict. This is incorrect, the value should be `(bool, date, datetime, dict, float, int, list, str, none_type)`
But because this is rendering `{{additionalPropertiesType}}` we would see the dict definition here which is confusing as this information is describing value types.
How would you fix this?

What do you want to do here?
```

Reply:
> How about changing the model name AdditionalPropertiesArray to AdditionalPropertiesArrayOfMapAnytype

The values of the undeclared (additional) properties must be an array, and the items in the array must be an object, they cannot be any type. So technically it's not "any type".
I'd rather not change the names in this PR because it's easier to make comparisons of the generated code for the issue raised in this PR.

> anytype_2: I think that this should this result in the data type `(bool, date, datetime, dict, float, int, list, str,)`.

I am going to try to understand the code a bit more.

> Why is this data type the same as anytype_1 + anytype_3?

It was already that way before, so from this perspective my PR is not disturbing whatever was implemented before. Let me go deeper into the code.

Reply:
The OAS schema for `anytype_1` is as below. That means the data type must be a dictionary; there are no declared properties, and since the `additionalProperties` keyword is not specified, there can be any undeclared properties and their value can be any type.
```yaml
      anytype_1:
        type: object
```

Before this PR, the generated data type was:
```
'anytype_1': (bool, date, datetime, dict, float, int, list, str,),
```

Isn't that incorrect? I.e. doesn't that mean that the value of `anytype_1` can be a boolean, a date, float, etc? But that's not correct, `anytype_1` cannot be a boolean, it must be a dict where the key is a string and the value can be anything.

Reply:
> How about changing the model name AdditionalPropertiesArray to AdditionalPropertiesArrayOfMapAnytype
> The values of the undeclared (additional) properties must be an array, and the items in the array must be an object, they cannot be any type. So technically it's not "any type".

I agree with you that's why I am suggesting a rename of:
AdditionalPropertiesArrayOfMapAnytype
Map is a synonym for dict/object. With the suffix `MapAnytype` I mean Map of AnyType, which means Map where the values can be any type. Is AdditionalPropertiesArrayOfMapOfAnytype clearer?

So I added these variables in a prior PR to give examples of AnyType and type: object usage. But my understanding at the time was incorrect. Now that we have a better  correct understanding here, why not make the values stored consistent with the variable names that describe those types?

Otherwise we are giving our users examples which are confusing and where the type described by the variable is different than what the actual type is.
Another solution here is to rename the current variables to be consistent with the updated types that are accepted. But those names get very long.
AdditionalPropertiesClass.map_array_anytype would be AdditionalPropertiesClass.map_map_array_anytype which feels less helpful to our users understanding because there are so many layers of types (4).

> anytype_1 is a bit of a misnomer
Yup, I agree, that's why I am asking that we rename those variables to:
- anytype_1 -> map_anytype1
- anytype_3 -> map_anytype1
Because we are storing dicts where the values can be any types.

The only case where the stored data type looks incorrect to me is anytype_2. All the other ones look like they were produced correcty.

Reply:
> So I added these variables in a prior PR to give examples of AnyType and type: object usage. But my understanding at the time was incorrect. Now that we have a better correct understanding here, why not make the values stored consistent with the variable names that describe those types?

Yes, I got that part from your earlier comment. I was saying I'd rather not make these changes in this PR because by keeping the names it's a lot easier to compare the changes introduced by this PR. The PR already has a lot of changes, it's getting tricky to compare all the changes.

> 
> Otherwise we are giving our users examples which are confusing and where the type described by the variable is different than what the actual type is.

Yes, I understand, but I'd rather not make these changes here. It's a separate improvement.

> Another solution here is to rename the current variables to be consistent with the updated types that are accepted. But those names get very long.
> AdditionalPropertiesClass.map_array_anytype would be AdditionalPropertiesClass.map_map_array_anytype which feels less helpful to our users understanding because there are so many layers of types (4).
> 
> > anytype_1 is a bit of a misnomer
> 
> Yup, I agree, that's why I am asking that we rename those variables to:
> 
> * anytype_1 -> map_anytype1
> * anytype_3 -> map_anytype2
>   Because we are storing dicts where the values can be any types.
> 
> The only case where the stored data type looks incorrect to me is anytype_2. All the other ones look like they were produced correcty.

Ok, let me work on fixing the generated data type for anytype_2, independant of the renaming.

Reply:
Okay, sounds good

Reply:
Ah, the problem with `anytype_2` is because the Schema.java instantiation is ObjectSchema with `"type: object"`, so it is incorrectly considered a property of type object. I am digging.

Reply:
Marking as resolved.